jxl/src/image/data_type.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+use std::fmt::Debug;
+
+mod private {
+    pub trait Sealed {}
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum DataTypeTag {
+    U8,
+    U16,
+    U32,
+    F32,
+    I8,
+    I16,
+    I32,
+    F16,
+    F64,
+}
+
+impl DataTypeTag {
+    pub const fn size(&self) -> usize {
+        match self {
+            DataTypeTag::U8 | DataTypeTag::I8 => 1,
+            DataTypeTag::U16 | DataTypeTag::F16 | DataTypeTag::I16 => 2,
+            DataTypeTag::U32 | DataTypeTag::F32 | DataTypeTag::I32 => 4,
+            DataTypeTag::F64 => 8,
+        }
+    }
+}
+
+const _: () = {
+    assert!(
+        DataTypeTag::I8
+            .size()
+            .is_multiple_of(std::mem::align_of::<i8>())
+    );
+    assert!(
+        DataTypeTag::U8
+            .size()
+            .is_multiple_of(std::mem::align_of::<u8>())
+    );
+    assert!(
+        DataTypeTag::I16
+            .size()
+            .is_multiple_of(std::mem::align_of::<i16>())
+    );
+    assert!(
+        DataTypeTag::U16
+            .size()
+            .is_multiple_of(std::mem::align_of::<u16>())
+    );
+    assert!(
+        DataTypeTag::F16
+            .size()
+            .is_multiple_of(std::mem::align_of::<half::f16>())
+    );
+    assert!(
+        DataTypeTag::I32
+            .size()
+            .is_multiple_of(std::mem::align_of::<i32>())
+    );
+    assert!(
+        DataTypeTag::U32
+            .size()
+            .is_multiple_of(std::mem::align_of::<u32>())
+    );
+    assert!(
+        DataTypeTag::F32
+            .size()
+            .is_multiple_of(std::mem::align_of::<f32>())
+    );
+    assert!(
+        DataTypeTag::F64
+            .size()
+            .is_multiple_of(std::mem::align_of::<f64>())
+    );
+    assert!(std::mem::size_of::<i8>() == DataTypeTag::I8.size());
+    assert!(std::mem::size_of::<u8>() == DataTypeTag::U8.size());
+    assert!(std::mem::size_of::<i16>() == DataTypeTag::I16.size());
+    assert!(std::mem::size_of::<u16>() == DataTypeTag::U16.size());
+    assert!(std::mem::size_of::<half::f16>() == DataTypeTag::F16.size());
+    assert!(std::mem::size_of::<i32>() == DataTypeTag::I32.size());
+    assert!(std::mem::size_of::<u32>() == DataTypeTag::U32.size());
+    assert!(std::mem::size_of::<f32>() == DataTypeTag::F32.size());
+    assert!(std::mem::size_of::<f64>() == DataTypeTag::F64.size());
+};
+
+/// # Safety
+/// Any type implementing this trait must be a "bag-of-bits" type with no padding.
```

Comment:
I wonder if using `bytemuck::Pod` would be one way to avoid 1) introducing an `unsafe trait` and 2) providing `unsafe impl`s for primitive types.

OTOH, `bytemuck::Pod` may not cover all of your requirements (matching `Self::DATA_TYPE_ID.size()`).  Still, maybe removal of _some_ of the safety requirements is desirable?

Reply:
I have thought about bytemuck, but I think that in this case (in which we just handle primitive types) the best tradeoff is to not add a dependency, given that (a) it is fairly simple to verify the safety requirements and (b) it wouldn't reduce code much.